cmd/dep/status.go

```diff
@@ -366,85 +408,157 @@ func runStatusAll(ctx *dep.Ctx, out outputter, p *dep.Project, sm gps.SourceMana
-				bs.Constraint = c.Constraint
-
-				vl, err := sm.ListVersions(proj.Ident())
-				if err == nil {
-					gps.SortPairedForUpgrade(vl)
-
-					for _, v := range vl {
-						// Because we've sorted the version list for
-						// upgrade, the first version we encounter that
-						// matches our constraint will be what we want.
-						if c.Constraint.Matches(v) {
-							bs.Latest = v.Revision()
-							break
+
+				// Only if we have a non-rev and non-plain version do/can we display
+				// anything wrt the version's updateability.
+				if bs.Version != nil && bs.Version.Type() != gps.IsVersion {
+					c, has := p.Manifest.Constraints[proj.Ident().ProjectRoot]
+					if !has {
+						c.Constraint = gps.Any()
+					}
+					// TODO: This constraint is only the constraint imposed by the
+					// current project, not by any transitive deps. As a result,
+					// transitive project deps will always show "any" here.
+					bs.Constraint = c.Constraint
+
+					vl, err := sm.ListVersions(proj.Ident())
+					if err == nil {
+						gps.SortPairedForUpgrade(vl)
+
+						for _, v := range vl {
+							// Because we've sorted the version list for
+							// upgrade, the first version we encounter that
+							// matches our constraint will be what we want.
+							if c.Constraint.Matches(v) {
+								bs.Latest = v.Revision()
+								break
+							}
 						}
+					} else {
+						// Failed to fetch version list (could happen due to
+						// network issue)
+						bs.hasError = true
+						errListVerCh <- err
 					}
 				}
-			}
 
-			out.BasicLine(&bs)
+				bsCh <- &bs
+
+			}(proj)
 		}
+
+		wg.Wait()
+		close(bsCh)
+		close(errListPkgCh)
+		close(errListVerCh)
+
+		// Newline after printing the status progress output
 		logger.Println()
+
+		// List Packages errors. This would happen only for dot output.
+		if len(errListPkgCh) > 0 {
+			err = errFailedListPkg
+			if ctx.Verbose {
+				for err := range errListPkgCh {
+					ctx.Err.Println(err.Error())
+				}
+				ctx.Err.Println()
+			}
+		}
+
+		// List Version errors
+		if len(errListVerCh) > 0 {
+			if err == nil {
+				err = errFailedUpdate
+			} else {
+				err = errMultipleFailures
+			}
+
+			// Count ListVersions error because we get partial results when
+			// this happens.
+			errCount = len(errListVerCh)
+			if ctx.Verbose {
+				for err := range errListVerCh {
+					ctx.Err.Println(err.Error())
+				}
+				ctx.Err.Println()
+			}
+		}
+
+		// A map of ProjectRoot and *BasicStatus. This is used in maintain the
+		// order of BasicStatus in output by collecting all the BasicStatus and
+		// then using them in order.
+		bsMap := make(map[string]*BasicStatus)
+		for bs := range bsCh {
+			bsMap[bs.ProjectRoot] = bs
+		}
+
+		// Use the collected BasicStatus in outputter
+		for _, proj := range slp {
+			out.BasicLine(bsMap[string(proj.Ident().ProjectRoot)])
+		}
+
 		out.BasicFooter()
 
-		return digestMismatch, hasMissingPkgs, nil
+		return false, false, errCount, err
```

Comment:
Can this be `false, false, 0, nil`?

Everything else looks good!

Reply:
err... no, I don't think so. That's the only place where we are counting and returning error, for partial results.

Reply:
Whoops thought I was looking at the final return.